oasis-reference/src/main/kotlin/com/empowerops/volition/ref_oasis/OptimizerController.kt

```diff
@@ -0,0 +1,230 @@
+package com.empowerops.volition.ref_oasis
+
+import javafx.beans.Observable
+import javafx.beans.property.SimpleStringProperty
+import javafx.collections.ObservableList
+import javafx.fxml.FXML
+import javafx.scene.control.*
+import javafx.scene.control.cell.TreeItemPropertyValueFactory
+import javafx.scene.input.KeyCode
+import javafx.scene.layout.AnchorPane
+import javafx.scene.layout.VBox
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.GlobalScope
+import kotlinx.coroutines.javafx.JavaFx
+import kotlinx.coroutines.launch
+import tornadofx.selectedItem
+import java.lang.NumberFormatException
+import java.time.Duration
+
+class ViewData(
+        val nodes: ObservableList<String>,
+        val allMessages: ObservableList<OptimizerEndpoint.Message>,
+        val currentEvaluationStatus: SimpleStringProperty,
+        val resultList: ObservableList<OptimizerEndpoint.Result>,
+        val updateList : ObservableList<String>){
```

Comment:
so, because the `Optimizer` depends on this class, and this class depends on `javafx.ObservableList`, our optimizer cant function without this particular UI framework.

Any ideas on how we might cut that dependency?

Reply:
added couple interfaces
I extract the model (view data, map used Endpoint) and put all the `update data` then `update view `  call as one action and expose those method via the Service interface. The service now do the job of `Endpoint: request a add node` -> `Service: add the node and update view model` -> `UI: update via binding`

For the other way around, OptimzierControl are the interface of endpoint in UI call for start, stop etc.